src/frontend/contexts/SettingsStoreContext.tsx

```diff
@@ -0,0 +1,170 @@
+import {createStore, useStore, type StoreApi} from 'zustand';
+import {
+  createJSONStorage,
+  persist as createPersistedState,
+} from 'zustand/middleware';
+import {useShallow} from 'zustand/react/shallow';
+
+import {MMKVZustandStorage} from '../hooks/persistedState/createPersistedState';
+import {CoordinateFormatSchema} from '../lib/coordinateFormat';
+
+export const SettingsStateSchema = v.object({
+  coordinateFormat: v.optional(CoordinateFormatSchema),
+  locale: v.optional(
+    v.union([
+      v.object({
+        languageTag: v.string(),
+      }),
+      v.literal('system'),
+    ]),
+  ),
+  manualCoordinateEntryFormat: v.optional(CoordinateFormatSchema),
+  metricsDiagnosticsPermissionsEnabled: v.optional(v.boolean()),
+});
+
+/**
+ * "Raw" settings state
+ */
+export type Settings = v.InferOutput<typeof SettingsStateSchema>;
+
+/**
+ * Settings fields that use a default value if missing
+ */
+type SettingWithDefault =
+  | 'coordinateFormat'
+  | 'locale'
+  | 'manualCoordinateEntryFormat'
+  | 'metricsDiagnosticsPermissionsEnabled';
+
+/**
+ * Settings with defaults filled in
+ */
+export type ResolvedSettings = SimplifyDeep<
+  SetNonNullable<Settings, SettingWithDefault>
+>;
+
+const SETTINGS_DEFAULTS = {
+  coordinateFormat: 'utm',
+  locale: 'system',
+  manualCoordinateEntryFormat: 'utm',
+  metricsDiagnosticsPermissionsEnabled: true,
+} as const satisfies Pick<ResolvedSettings, SettingWithDefault>;
+
+// NOTE: Do not change!
+export const STORAGE_KEY = 'Settings' as const;
+
+function createInitialState(): Settings {
+  return {};
+}
+
+export function createSettingsStore({persist} = {persist: false}) {
+  let store: StoreApi<Settings>;
+
+  if (persist) {
+    store = createStore(
+      createPersistedState(createInitialState, {
+        name: STORAGE_KEY,
+        storage: createJSONStorage(() => MMKVZustandStorage),
+        version: 1,
+        migrate: (persistedState, version): Settings => {
+          const migratedState = createInitialState();
+
+          if (version === 0) {
+            if (typeof persistedState !== 'object') return migratedState;
+            if (persistedState === null) return migratedState;
+
+            if (
+              'coordinateFormat' in persistedState &&
+              v.is(
+                SettingsStateSchema.entries['coordinateFormat'],
+                persistedState.coordinateFormat,
+              )
+            ) {
+              migratedState.coordinateFormat = persistedState.coordinateFormat;
+            }
+
+            if (
+              'manualCoordinateEntryFormat' in persistedState &&
+              v.is(
+                SettingsStateSchema.entries['manualCoordinateEntryFormat'],
+                persistedState.manualCoordinateEntryFormat,
+              )
+            ) {
+              migratedState.manualCoordinateEntryFormat =
+                persistedState.manualCoordinateEntryFormat;
+            }
+          }
+
+          return migratedState;
+        },
```

Comment:
> I also recommend using the valibot schema in the migration function, and not setting the value (e.g. reverting to defaults) if an invalid value is found, in case we mess up migration somehow.

@gmaclennan can you elaborate on this? As currently implemented, this migration function does the following:

- ports the existing `coordinateFormat` from v0 settings
- ports the existing `manualCoordinateEntryFormat` from v0 settings

for the other fields in v1, it applies the initial state i.e. omitting the field entirely (since initial settings state is now `{}`).